In [1]:
#Import libraries
from pyforest import *
lazy_imports()
import warnings
warnings.filterwarnings('ignore')

In [2]:
#import the dataset into pandas: 
data = pd.ExcelFile('Assignment.xlsx', engine = 'openpyxl')

<IPython.core.display.Javascript object>

In [3]:
#Customers Monthly dataset
df_cust = pd.read_excel(data, 'Customers Monthly')
df_cust = df_cust[df_cust['CUSTOMER_HK'].notna()]
df_cust

<IPython.core.display.Javascript object>

CUSTOMER_HK     DAY_DT  \
0      04c604e678b18d7ca10b85fb8d3672a03b7b6b71e78dd5... 2020-11-30   
1      a42a9fdf0ee31d331c9d96ce6fbf30f307b52ff40a55ec... 2021-01-31   
2      be92b2698f17aa7fab796fd5eb132cc251fce6fe1870dc... 2020-12-31   
3      c84455bb3bbc2dbf16a1edd9c44e76baaf06b0ee224ed5... 2020-12-31   
4      7df3aa4e5f29e4f5946d652d4f6d6f88d3f850ac9ba33a... 2020-12-31   
...                                                  ...        ...   
39530  4c65aeb4e3c56d9ad692a2dd53f6e58d3b1c03f53ebee8... 2021-01-31   
39531  b3c4de44a98d55acc296d54636f321323e55bc793f0c01... 2020-12-31   
39532  ed1cd862b1e0c82f9c96b6a895c6122fbfcac469801701... 2020-12-31   
39533  68b47f0f4243f0fccc0e21bc72b750f76387838e20aa82... 2020-11-30   
39534  12a2683372a4a0a3ffaeed2fd17bd789e14893bcb094ed... 2020-11-30   

       BANK_IS_ACTIVE  INVEST_IS_ACTIVE PLAN_NM  
0                 1.0               0.0   prime  
1                 1.0               0.0   prime  
2                 1.0               0.0   prime  
3                 1.0               0.0   prime  
4                 0.0               0.0   prime  
...               ...               ...     ...  
39530             1.0               0.0   prime  
39531             1.0               0.0   prime  
39532             1.0               0.0   prime  
39533             1.0               0.0   prime  
39534             0.0               0.0   prime  

[39535 rows x 5 columns]

In [4]:
# Handling Missing Values
num_missing = df_cust.isnull().sum()
missing_per = (num_missing/df_cust.shape[0])*100
missing_table = pd.concat([num_missing,missing_per], axis=1, ignore_index=True) 
missing_table.rename(columns={0:'Num_Missing',1:'%_Missing'}, inplace=True)
print('\033[1m' + '\033[94m' + 'Missing Data Info:'.center(80) + '\033[0m')
print(missing_table)

<IPython.core.display.Javascript object>

                               Missing Data Info:                               
                  Num_Missing  %_Missing
CUSTOMER_HK                 0   0.000000
DAY_DT                      0   0.000000
BANK_IS_ACTIVE              0   0.000000
INVEST_IS_ACTIVE            0   0.000000
PLAN_NM                  1173   2.966991


In [5]:
# Only ca. 3% of PLAN_NM is missing so we will remove those
df_cust.dropna(subset=['PLAN_NM'], inplace=True)
df_cust

CUSTOMER_HK     DAY_DT  \
0      04c604e678b18d7ca10b85fb8d3672a03b7b6b71e78dd5... 2020-11-30   
1      a42a9fdf0ee31d331c9d96ce6fbf30f307b52ff40a55ec... 2021-01-31   
2      be92b2698f17aa7fab796fd5eb132cc251fce6fe1870dc... 2020-12-31   
3      c84455bb3bbc2dbf16a1edd9c44e76baaf06b0ee224ed5... 2020-12-31   
4      7df3aa4e5f29e4f5946d652d4f6d6f88d3f850ac9ba33a... 2020-12-31   
...                                                  ...        ...   
39530  4c65aeb4e3c56d9ad692a2dd53f6e58d3b1c03f53ebee8... 2021-01-31   
39531  b3c4de44a98d55acc296d54636f321323e55bc793f0c01... 2020-12-31   
39532  ed1cd862b1e0c82f9c96b6a895c6122fbfcac469801701... 2020-12-31   
39533  68b47f0f4243f0fccc0e21bc72b750f76387838e20aa82... 2020-11-30   
39534  12a2683372a4a0a3ffaeed2fd17bd789e14893bcb094ed... 2020-11-30   

       BANK_IS_ACTIVE  INVEST_IS_ACTIVE PLAN_NM  
0                 1.0               0.0   prime  
1                 1.0               0.0   prime  
2                 1.0               0.0   prime  
3                 1.0               0.0   prime  
4                 0.0               0.0   prime  
...               ...               ...     ...  
39530             1.0               0.0   prime  
39531             1.0               0.0   prime  
39532             1.0               0.0   prime  
39533             1.0               0.0   prime  
39534             0.0               0.0   prime  

[38362 rows x 5 columns]

In [6]:
#For Conversations dataset

df_con = pd.read_excel(data, 'Conversations')
df_con = df_con[df_con['CUSTOMER_HK'].notna()] #remove rows with missing values in CUSTOMER_HK
df_con

<IPython.core.display.Javascript object>

CONVERSATION_HK          CREATED_AT  \
0      0008a84b8d7f6bf55740053ca635b11722bfe4f38a9ffa... 2020-11-16 17:18:40   
1      000ad81c0e996b624c296cb105b11c5f76777fae456ecb... 2020-12-08 09:13:49   
2      000c2e8ad3fb870813d3ea8f7ea1094f7508d9fee1205f... 2020-12-28 14:09:31   
3      000cc854daaeff5b59deaae9f11c18c0d59beda58b7249... 2021-01-08 08:49:28   
4      000db3ed6e7c9a1ab900a61fa02f2fc6d75fb65670d6b0... 2020-12-03 22:15:15   
...                                                  ...                 ...   
40310  fffa6fcad633f39649228bdd99b6f1fc2365067e6fd0d6... 2021-01-04 15:56:03   
40311  fffb4d2846b9fed3be3d6b6ac195057e57bf34a3650d4b... 2020-12-23 13:32:14   
40312  fffcd914a96f3b7d2468cdd704d3857cc12ebfb12baf9b... 2020-11-20 18:24:26   
40313  fffda5c392e3c25ff2eac6ff3a1f6c8cd0838278490eaa... 2020-12-07 05:41:18   
40314  fffee1f9d3b3304a99ef46fa7322d77a5d88792b91e032... 2020-12-10 13:37:44   

      STATUS_CD           CLOSED_AT  RATING_NUM LANGUAGE_CD  FRT_DU_IN_MIN  \
0        closed 2020-11-16 18:41:31         NaN          de      82.816667   
1        closed 2020-12-13 09:04:20         NaN          de    5759.333333   
2        closed 2020-12-28 15:58:05         NaN          fr      35.666667   
3        closed 2021-01-13 08:09:37         NaN          de     653.800000   
4        closed 2020-12-05 01:58:26         NaN          en     843.033333   
...         ...                 ...         ...         ...            ...   
40310    closed 2021-01-04 20:17:36         NaN          en     260.566667   
40311    closed 2020-12-23 15:19:54         4.0          fr      27.033333   
40312    closed 2020-11-23 10:38:23         NaN          en    1772.050000   
40313    closed 2020-12-07 06:07:21         NaN          en       2.933333   
40314    closed                 NaT         NaN          de            NaN   

      INITIATOR_CD                                        CUSTOMER_HK  
0         Customer  04c604e678b18d7ca10b85fb8d3672a03b7b6b71e78dd5...  
1         Customer  a42a9fdf0ee31d331c9d96ce6fbf30f307b52ff40a55ec...  
2         Customer  be92b2698f17aa7fab796fd5eb132cc251fce6fe1870dc...  
3         Customer  c84455bb3bbc2dbf16a1edd9c44e76baaf06b0ee224ed5...  
4         Customer  7df3aa4e5f29e4f5946d652d4f6d6f88d3f850ac9ba33a...  
...            ...                                                ...  
40310     Customer  4c65aeb4e3c56d9ad692a2dd53f6e58d3b1c03f53ebee8...  
40311     Customer  b3c4de44a98d55acc296d54636f321323e55bc793f0c01...  
40312     Customer  ed1cd862b1e0c82f9c96b6a895c6122fbfcac469801701...  
40313     Customer  68b47f0f4243f0fccc0e21bc72b750f76387838e20aa82...  
40314  Vivid Money  12a2683372a4a0a3ffaeed2fd17bd789e14893bcb094ed...  

[40315 rows x 9 columns]

In [7]:
# Data Transformation

## Handling Missing Values
num_missing = df_con.isnull().sum()
missing_per = (num_missing/df_con.shape[0])*100
missing_table = pd.concat([num_missing,missing_per], axis=1, ignore_index=True) 
missing_table.rename(columns={0:'Num_Missing',1:'%_Missing'}, inplace=True)
print('\033[1m' + '\033[94m' + 'Missing Data Info:'.center(80) + '\033[0m')
print(missing_table)

<IPython.core.display.Javascript object>

                               Missing Data Info:                               
                 Num_Missing  %_Missing
CONVERSATION_HK            0   0.000000
CREATED_AT                 0   0.000000
STATUS_CD                  0   0.000000
CLOSED_AT                756   1.875233
RATING_NUM             25230  62.582165
LANGUAGE_CD                2   0.004961
FRT_DU_IN_MIN           6264  15.537641
INITIATOR_CD               0   0.000000
CUSTOMER_HK                0   0.000000


In [8]:
#convert to datetime format 
df_con['CREATED_AT'] = pd.to_datetime(df_con['CREATED_AT'].astype(str), format='%Y-%m-%d %H:%M:%S')
df_con['YEAR'] = df_con['CREATED_AT'].apply(lambda x: x.year)
df_con['CREATED_DATE'] = df_con['CREATED_AT'].dt.date
df_con['CREATED_YEAR'] = df_con['CREATED_AT'].dt.year
df_con['CREATED_MONTH'] = df_con['CREATED_AT'].dt.month
df_con['CREATED_MONTH_NAME'] = df_con['CREATED_AT'].dt.month_name()
df_con['CREATED_YEAR_MONTH'] = df_con['CREATED_AT'].dt.strftime('%Y-%m')
df_con['CREATED_DAY'] = df_con['CREATED_AT'].dt.day
df_con['CREATED_DAY_NAME'] = df_con['CREATED_AT'].dt.day_name()
df_con['CREATED_HOUR'] = df_con['CREATED_AT'].dt.hour
#df_con.set_index('date', inplace=True)

<IPython.core.display.Javascript object>

In [9]:
#drop rows with NaNs in CLOSED_AT & FRT_DU_IN_MIN columns
df_conv = df_con.dropna(subset=['CLOSED_AT','FRT_DU_IN_MIN'])

In [10]:
##Chat Handle Time (CHT)

# Create new column with calculated Chat Handle Time in minutes (CHT)
df_conv['CHT_MIN'] = df_conv['CLOSED_AT'] - df_conv['CREATED_AT']
df_conv['CHT_MIN']=df_conv['CHT_MIN']/np.timedelta64(1,'m')

<IPython.core.display.Javascript object>

In [11]:
df_conv.columns

Index(['CONVERSATION_HK', 'CREATED_AT', 'STATUS_CD', 'CLOSED_AT', 'RATING_NUM',
       'LANGUAGE_CD', 'FRT_DU_IN_MIN', 'INITIATOR_CD', 'CUSTOMER_HK', 'YEAR',
       'CREATED_DATE', 'CREATED_YEAR', 'CREATED_MONTH', 'CREATED_MONTH_NAME',
       'CREATED_YEAR_MONTH', 'CREATED_DAY', 'CREATED_DAY_NAME', 'CREATED_HOUR',
       'CHT_MIN'],
      dtype='object')

In [12]:
#rearrange columns
df_conv = df_conv[['CUSTOMER_HK', 'CONVERSATION_HK', 'INITIATOR_CD', 'LANGUAGE_CD', 'CREATED_AT', 'CREATED_DATE', 
                   'CREATED_YEAR', 'CREATED_YEAR_MONTH', 'CREATED_MONTH', 'CREATED_MONTH_NAME', 'CREATED_DAY', 
                   'CREATED_DAY_NAME', 'CREATED_HOUR', 'CLOSED_AT', 'FRT_DU_IN_MIN', 'CHT_MIN', 'STATUS_CD', 'RATING_NUM']]

In [13]:
df_conv

CUSTOMER_HK  \
0      04c604e678b18d7ca10b85fb8d3672a03b7b6b71e78dd5...   
1      a42a9fdf0ee31d331c9d96ce6fbf30f307b52ff40a55ec...   
2      be92b2698f17aa7fab796fd5eb132cc251fce6fe1870dc...   
3      c84455bb3bbc2dbf16a1edd9c44e76baaf06b0ee224ed5...   
4      7df3aa4e5f29e4f5946d652d4f6d6f88d3f850ac9ba33a...   
...                                                  ...   
40309  d0c5852da05dd8caa02f4c276659abcdf4da5a285ed363...   
40310  4c65aeb4e3c56d9ad692a2dd53f6e58d3b1c03f53ebee8...   
40311  b3c4de44a98d55acc296d54636f321323e55bc793f0c01...   
40312  ed1cd862b1e0c82f9c96b6a895c6122fbfcac469801701...   
40313  68b47f0f4243f0fccc0e21bc72b750f76387838e20aa82...   

                                         CONVERSATION_HK INITIATOR_CD  \
0      0008a84b8d7f6bf55740053ca635b11722bfe4f38a9ffa...     Customer   
1      000ad81c0e996b624c296cb105b11c5f76777fae456ecb...     Customer   
2      000c2e8ad3fb870813d3ea8f7ea1094f7508d9fee1205f...     Customer   
3      000cc854daaeff5b59deaae9f11c18c0d59beda58b7249...     Customer   
4      000db3ed6e7c9a1ab900a61fa02f2fc6d75fb65670d6b0...     Customer   
...                                                  ...          ...   
40309  fff88b887df6898987707eb867244d7792bc9cff1af8c5...     Customer   
40310  fffa6fcad633f39649228bdd99b6f1fc2365067e6fd0d6...     Customer   
40311  fffb4d2846b9fed3be3d6b6ac195057e57bf34a3650d4b...     Customer   
40312  fffcd914a96f3b7d2468cdd704d3857cc12ebfb12baf9b...     Customer   
40313  fffda5c392e3c25ff2eac6ff3a1f6c8cd0838278490eaa...     Customer   

      LANGUAGE_CD          CREATED_AT CREATED_DATE  CREATED_YEAR  \
0              de 2020-11-16 17:18:40   2020-11-16          2020   
1              de 2020-12-08 09:13:49   2020-12-08          2020   
2              fr 2020-12-28 14:09:31   2020-12-28          2020   
3              de 2021-01-08 08:49:28   2021-01-08          2021   
4              en 2020-12-03 22:15:15   2020-12-03          2020   
...           ...                 ...          ...           ...   
40309          en 2021-01-12 07:49:58   2021-01-12          2021   
40310          en 2021-01-04 15:56:03   2021-01-04          2021   
40311          fr 2020-12-23 13:32:14   2020-12-23          2020   
40312          en 2020-11-20 18:24:26   2020-11-20          2020   
40313          en 2020-12-07 05:41:18   2020-12-07          2020   

      CREATED_YEAR_MONTH  CREATED_MONTH CREATED_MONTH_NAME  CREATED_DAY  \
0                2020-11             11           November           16   
1                2020-12             12           December            8   
2                2020-12             12           December           28   
3                2021-01              1            January            8   
4                2020-12             12           December            3   
...                  ...            ...                ...          ...   
40309            2021-01              1            January           12   
40310            2021-01              1            January            4   
40311            2020-12             12           December           23   
40312            2020-11             11           November           20   
40313            2020-12             12           December            7   

      CREATED_DAY_NAME  CREATED_HOUR           CLOSED_AT  FRT_DU_IN_MIN  \
0               Monday            17 2020-11-16 18:41:31      82.816667   
1              Tuesday             9 2020-12-13 09:04:20    5759.333333   
2               Monday            14 2020-12-28 15:58:05      35.666667   
3               Friday             8 2021-01-13 08:09:37     653.800000   
4             Thursday            22 2020-12-05 01:58:26     843.033333   
...                ...           ...                 ...            ...   
40309          Tuesday             7 2021-01-13 14:38:42    1848.300000   
40310           Monday            15 2021-01-04 20:17:36     260.566667   
40311        Wednesday            13 2020-12-23 15:19:

In [14]:
#save to csv for visualization reporting in your BI tool
df_conv.to_csv('conversations_BI.csv')
df_cust.to_csv('customers_BI.csv')

In [15]:
# Lets check the summary statistics before & after transformation of the variables 

##summary statistics before transformation
df_con.describe(include='all')

CONVERSATION_HK  \
count                                               40315   
unique                                              39535   
top     af085d8d54688cbac9eff40aeddea2a448ff14f87e755e...   
freq                                                    2   
first                                                 NaN   
last                                                  NaN   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                 CREATED_AT STATUS_CD            CLOSED_AT    RATING_NUM  \
count                 40315     40315                39559  15085.000000   
unique                39367         3                38650           NaN   
top     2021-01-04 15:46:41    closed  2021-01-04 17:42:02           NaN   
freq                      4     40303                    3           NaN   
first   2020-11-01 00:44:03       NaN  2020-11-01 01:21:50           NaN   
last    2021-01-31 23:51:57       NaN  2021-10-12 09:24:58           NaN   
mean                    NaN       NaN                  NaN      4.201724   
std                     NaN       NaN                  NaN      1.158345   
min                     NaN       NaN                  NaN      1.000000   
25%                     NaN       NaN                  NaN      4.000000   
50%                     NaN       NaN                  NaN      5.000000   
75%                     NaN       NaN                  NaN      5.000000   
max                     NaN       NaN                  NaN      5.000000   

       LANGUAGE_CD  FRT_DU_IN_MIN INITIATOR_CD  \
count        40313   34051.000000        40315   
unique           5            NaN            2   
top             de            NaN     Customer   
freq         15233            NaN        38071   
first          NaN            NaN          NaN   
last           NaN            NaN          NaN   
mean           NaN     838.680738          NaN   
std            NaN    3640.505013          NaN   
min            NaN       0.183333          NaN   
25%            NaN       9.333333          NaN   
50%            NaN      59.783333          NaN   
75%            NaN     773.616667          NaN   
max            NaN  372641.100000          NaN   

                                              CUSTOMER_HK          YEAR  \
count                                               40315  40315.000000   
unique                                              19717           NaN   
top     238fe4ef9aa063a4ffe2424a20ead5ab9d269cf43ac328...           NaN   
freq                                                   70           NaN   
first                                                 NaN           NaN   
last                                                  NaN           NaN   
mean                                                  NaN   2020.375245   
std                                                   NaN      0.484192   
min                                                   NaN   2020.000000   
25%                                                   NaN   2020.000000   
50%                                                   NaN   2020.000000   
75%                                                   NaN   2021.000000   
max                                                   NaN   2021.000000   

       CREATED_DATE  CREATED_YEAR  CREATED_MONTH CREATED_MONTH_NAME  \
count         40315  40315.000000   40315.000000              40315   
unique           92           NaN            NaN                  3   
top      2020-11-18           NaN            NaN            January   
freq            797           NaN            NaN              15128   
first          

In [16]:
# summary statistics after transformation
df_conv.describe(include='all')

CUSTOMER_HK  \
count                                               34044   
unique                                              18025   
top     238fe4ef9aa063a4ffe2424a20ead5ab9d269cf43ac328...   
freq                                                   48   
first                                                 NaN   
last                                                  NaN   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                                          CONVERSATION_HK INITIATOR_CD  \
count                                               34044        34044   
unique                                              33367            1   
top     aefff4d19ae9e3102994a69923e3f9cced94c051e88a8a...     Customer   
freq                                                    2        34044   
first                                                 NaN          NaN   
last                                                  NaN          NaN   
mean                                                  NaN          NaN   
std                                                   NaN          NaN   
min                                                   NaN          NaN   
25%                                                   NaN          NaN   
50%                                                   NaN          NaN   
75%                                                   NaN          NaN   
max                                                   NaN          NaN   

       LANGUAGE_CD           CREATED_AT CREATED_DATE  CREATED_YEAR  \
count        34044                34044        34044  34044.000000   
unique           5                33253           92           NaN   
top             de  2021-01-04 15:46:41   2020-11-18           NaN   
freq         13104                    4          687           NaN   
first          NaN  2020-11-01 00:44:03          NaN           NaN   
last           NaN  2021-01-31 23:51:57          NaN           NaN   
mean           NaN                  NaN          NaN   2020.390289   
std            NaN                  NaN          NaN      0.487822   
min            NaN                  NaN          NaN   2020.000000   
25%            NaN                  NaN          NaN   2020.000000   
50%            NaN                  NaN          NaN   2020.000000   
75%            NaN                  NaN          NaN   2021.000000   
max            NaN                  NaN          NaN   2021.000000   

       CREATED_YEAR_MONTH  CREATED_MONTH CREATED_MONTH_NAME   CREATED_DAY  \
count               34044   34044.000000              34044  34044.000000   
unique                  3            NaN                  3           NaN   
top               2021-01            NaN            January           NaN   
freq                13287            NaN              13287           NaN   
first                 NaN            NaN                NaN           NaN   
last                  NaN            NaN                NaN           NaN   
mean                  NaN       7.431207                NaN     16.265832   
std                   NaN       5.160184                NaN      8.766111   
min                   NaN       1.000000                NaN      1.000000   
25%                   NaN       1.000000                NaN      9.000000   
50%                   NaN      11.000000                NaN     16.000000   
75%                   NaN      12.000000                NaN     24.000000   
max                   NaN      12.000000                NaN     31.000000   

       CREATED_DAY_NAME  CREATED_HOUR            CLOSED_AT  FRT_DU_IN_MIN  \
count            

Some important observations after removing ca. 16% ~ 6200 rows:

* For CUSTOMER_HK, there are now 18025 unique customers, down from 19717	
* For CONVERSATION_HK, there are now 33367 unique conversation keys, down from 39535. With a total chat count of 34044, it means only 677 ~ 2% chats required follow up.
* For INITIATOR_CD, there is only converations initiated by customers, with those from Vivid money agents dropped.
* For STATUS_CD, all open cases were dropped with only closed & snoozed cases retained.
* For FRT_DU_IN_MIN, the Average, Minimum & Maximum all remained ca. 838 mins (14 hrs), 0.18 mins, & 372641 mins (259 days) respectively. 25% of the chats had FRT < 9.33 mins, half had FRT < 1 hour & 75% had FRT < 13hrs. 
* For CHT_MIN, the Average, Minimum & Maximum of ca. 7694 mins (5.4 days), 0.55 mins, & 459374 mins (319 days) respectively. 25% of the chats had CHT < 1.5 hrs, half had CHT < 14 hrs & 75% had CHT < 3 days. 
* After removing rows with missing values in closed_at and rating_num columns, ca. 15%, we find out that Open is missing from Status_CD, Vivid Money is missing from Initiator_CD, which might imply that most and or all chats initiated by vivid money agents were open and or not rated. We can look further into this.